### PR TITLE
Add environment.yml file for conda installation

### PR DESCRIPTION
**Pull request by donerancl**
This pull request will add an `environment.yml` file for easy conda installation
The following was successful for me...
```conda env create -f environment.yml```
```pip install .```

### CODE REVIEW

**Comment by kevingreenman** [comment]
It looks like you generated this using the `--no-builds` argument when doing the `conda env export`, right? I agree that's a good way to do it to make it more platform- and machine-independent. 

I was also going to suggest that we use the `--from-history` argument when doing the `conda env export`, but I realized that since all the packages in the initial environment were installed using pip, that argument might not work. Maybe you already tried that.

 If we want to have a simpler a less fragile env file more similar to the one from v1, we may have to do the following:

1. create a new blank conda env with python 3.11
2. use conda to manually install each of the packages with the required version from `pyproject.toml`
3. export the conda env with the `--no-builds` and `--from-history` arguments
4. use the exported yml file to create a new environment
5. manually install pytest and pytest-cov in the new environment
6. run `pytest -v` to verify that all the tests pass using the clean environment

**Comment by donerancl** [comment]
@kevingreenman I followed the steps you gave, and my `environment.yml` file has much fewer dependencies, but there are still many additional ones not listed in the `pyproject.toml` file. I have the tests running now, so far all successful.

```
name: chemprop
channels:
  - conda-forge
dependencies:
  - readline=8.2
  - debugpy=1.8.1
  - jupyter_core=5.7.2
  - matplotlib-inline=0.1.6
  - jupyter_client=8.6.1
  - libgcc-ng=13.2.0
  - setuptools=69.2.0
  - bzip2=1.0.8
  - pygments=2.17.2
  - pyzmq=25.1.2
  - libgomp=13.2.0
  - zipp=3.17.0
  - libsodium=1.0.18
  - libuuid=2.38.1
  - importlib_metadata=7.1.0
  - ncurses=6.4.20240210
  - libnsl=2.0.1
  - libstdcxx-ng=13.2.0
  - pip=24.0
  - pandas
  - platformdirs=4.2.0
  - tk=8.6.13
  - openssl
  - python=3.11.8
  - zeromq=4.3.5
  - executing=2.0.1
  - ca-certificates
  - pexpect=4.9.0
  - psutil=5.9.8
  - lightning
  - wcwidth=0.2.13
  - prompt-toolkit=3.0.42
  - scikit-learn
  - six=1.16.0
  - xz=5.2.6
  - pickleshare=0.7.5
  - astartes
  - ipykernel=6.29.3
  - ld_impl_linux-64=2.40
  - libsqlite=3.45.2
  - comm=0.2.2
  - _openmp_mutex=4.5
  - libzlib=1.2.13
  - ptyprocess=0.7.0
  - pure_eval=0.2.2
  - python_abi=3.11
  - decorator=5.1.1
  - nest-asyncio=1.6.0
  - wheel=0.43.0
  - asttokens=2.4.1
  - traitlets=5.14.2
  - parso=0.8.4
  - tornado=6.4
  - importlib-metadata=7.1.0
  - libexpat=2.6.2
  - jedi=0.19.1
  - stack_data=0.6.2
  - _libgcc_mutex=0.1
  - libffi=3.4.2
  - packaging=24.0
  - libxcrypt=4.4.36
  - rdkit
  - typing_extensions=4.11.0
  - exceptiongroup=1.2.0
  - ipython=8.22.2
  - pytorch
  - torchaudio
  - pytorch-cuda=11.8
  - torchvision
prefix: /home/donera/mambaforge/envs/chemprop
```

**Comment by kevingreenman** [comment]
@donerancl okay, yeah that definitely looks better. I don't think there's really a perfect way to do this, nor does it need to be perfect since this isn't our recommended installation method. So I think if you just remove the `prefix: /home/donera/mambaforge/envs/chemprop` line from the bottom and all the tests pass, then we can move forward with this. 

We may need to add some additional dependencies to this file once the hyperparameter optimization PR is merged (namely, `ray[tune]` and `hyperopt`). Although those are specified as an optional dependency set in the`pyproject.toml` in #734, so one could argue that we don't need to include them in the `environment.yml` (or if we did include those, we should also include the optional dependencies for testing, docs, etc.).

**Comment by donerancl** [comment]
I got an even shorter `environment.yml` file by repeating the process. I'm not sure what was different. Two tests failed, due to the astartes installation I guess. Does it need to be installed with pip?

```
FAILED tests/unit/data/test_data_utils.py::test_kennard_stone[mol_data0] - astartes.utils.exceptions.MoleculesNotInstalledError: To use molecule featurizer, install astartes with pip install astartes[molecules] or conda install ...
FAILED tests/unit/data/test_data_utils.py::test_kmeans[mol_data0] - astartes.utils.exceptions.MoleculesNotInstalledError: To use molecule featurizer, install astartes with pip install astartes[molecules] or conda install ...
```

File: environment.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+name: chemprop
+channels:
+  - pytorch
+  - conda-forge
+dependencies:
+  - python=3.11
```
Comment (kevingreenman) [inline environment.yml:6]:
```suggestion
  - python>=3.11
```

File: environment.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  - pytorch::pytorch
+  - pytorch::torchaudio
+  - pytorch::torchvision
+  - astartes[molecules]
+  - aimsim
+  - lightning
```
Comment (kevingreenman) [inline environment.yml:12]:
```suggestion
  - lightning>=2.0
```

File: environment.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  - numpy
```
Comment (kevingreenman) [inline environment.yml:13]:
```suggestion
  - numpy<2.0
```

Comment (kevingreenman) [reply environment.yml:13]: per #733 and #516

File: environment.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+name: chemprop
+channels:
+  - pytorch
+  - conda-forge
+dependencies:
+  - python=3.11
+  - pytorch::pytorch
```
Comment (kevingreenman) [inline environment.yml:7]: our `pyproject.toml` specifies `torch >= 2.1` [here](https://github.com/chemprop/chemprop/blob/ec55383d2888ae0a2d631d76df28c5fef3b7536c/pyproject.toml#L36). Can we specify that in this line? I'm not familiar with the `::` syntax, but maybe you can still just append `>=2.1` to the end of this line?

File: environment.yml (hunk 2 above)
Comment (kevingreenman) [reply environment.yml:12]: to match `pyproject.toml`

File: environment.yml (hunk 1 above)
Comment (kevingreenman) [reply environment.yml:6]: to match `pyproject.toml`

**Review by kevingreenman — CHANGES_REQUESTED** [review]
This looks much better! I left a few small suggestions where I think we should specify minimum/maximum versions in this file to match the `pyproject.toml`. 

@davidegraff also suggested restricting the versions of a few other packages in his comment [here](https://github.com/chemprop/chemprop/issues/499#issuecomment-1974938853), but I'm not sure what the reasoning for those was since we don't have those restrictions in the `pyproject.toml`.

@donerancl if you agree with my suggestions, could you add these restrictions and test it one more time?

File: environment.yml (hunk 4 above)
Comment (donerancl) [reply environment.yml:7]: yes, that should be fine!

**Comment by kevingreenman** [comment]
@donerancl thanks! If you can confirm that the tests all pass with these latest changes, I think we can go ahead and merge this

File: environment.yml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+name: chemprop
+channels:
+  - pytorch
+  - conda-forge
+dependencies:
+  - python>=3.11
+  - pytorch::pytorch>=2.1
+  - pytorch::torchaudio
+  - pytorch::torchvision
+  - astartes[molecules]
```
Comment (JacksonBurns) [inline environment.yml:10]:
```suggestion
  - astartes
```

The `astartes[molecules]` package is only for PyPI - to get it on `conda`, you just do `conda install astartes aimsim`, which we have accomplished here by specifying `aimsim` in the next line.

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
We need to add CI testing to this file to make sure that we don't forget about it in the future, as we did with v1. I'm going to tack on a quick commit for discussion.

**Comment by JacksonBurns** [comment]
@donerancl @kevingreenman I found that there was a missing dependency in the environment file, which I added in 1d439b6. The failure on 3.12 will be resolved shortly - I am releasing a new version of an upstream dependency with the problematic dependency removed. Re-running the CI some hours after this comment should resolve the issue.

**Comment by donerancl** [comment]
@kevingreenman @JacksonBurns Yes, it looks like my tests passed Friday evening. Did the missing dependency `configargparse` cause tests to fail?

**Comment by JacksonBurns** [comment]
On 3.12 the tests failed because of an upstream problem, which I have resolved. On 3.11 (and also 3.12 if not for the previous problem), the tests failed because that package was missing.

File: environment.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+name: chemprop
+channels:
+  - pytorch
+  - conda-forge
+dependencies:
+  - python>=3.11
+  - pytorch::pytorch>=2.1
+  - pytorch::torchaudio
+  - pytorch::torchvision
```
Comment (kevingreenman) [inline environment.yml:9]:
```suggestion
```

**Review by kevingreenman — APPROVED** [review]

**Pull request merged**